django/contrib/auth/management/__init__.py

```diff
@@ -111,47 +113,73 @@ def create_permissions(
         # models do not retain foreign key constraints.
         from django.contrib.auth.models import Permission
 
-        db = schema_editor.connection.alias
-        ctypes = ContentType.objects.using(db).filter(
-            app_label=self.app_label, model__icontains=old_model.lower()
+        perms = Permission.objects.using(self.using).filter(
+            content_type__app_label=self.app_label,
+            codename__in=[
+                action + "_" + from_model.lower()
+                for action in model_class._meta.default_permissions
+            ],
         )
-        for permission in Permission.objects.using(db).filter(
-            content_type_id__in=ctypes.values("id")
-        ):
-            prefix = permission.codename.split("_")[0]
-            default_verbose_name = camel_case_to_spaces(new_model)
-
-            new_codename = f"{prefix}_{new_model.lower()}"
-            new_name = f"Can {prefix} {default_verbose_name}"
-
-            if permission.codename != new_codename or permission.name != new_name:
-                permission.codename = new_codename
-                permission.name = new_name
+        for permission in perms:
+            action = permission.codename.split("_")[0]
+            new_codename = action + "_" + to_model.lower()
+            if self.backward:
+                # We don't have the target model class, so approximate.
+                # TODO: might be able to get this if we adjust the order of
+                # injected operations when migrating backwards.
+                verbose_name_raw = camel_case_to_spaces(from_model)
+            else:
+                verbose_name_raw = model_class._meta.verbose_name_raw
+            new_name = f"Can {action} {verbose_name_raw}"
+
+            if permission.codename != new_codename or permission.name != new_codename:
                 try:
-                    with transaction.atomic(using=db):
-                        permission.save(update_fields={"name", "codename"})
-                except IntegrityError:
-                    pass
-
-    def rename_forward(self, apps, schema_editor):
-        self._rename(apps, schema_editor, self.old_model, self.new_model)
-
-    def rename_backward(self, apps, schema_editor):
-        self._rename(apps, schema_editor, self.new_model, self.old_model)
+                    new_perm = Permission.objects.using(self.using).get(
+                        content_type_id=permission.content_type_id,
+                        codename=new_codename,
+                    )
+                except Permission.DoesNotExist:
+                    permission.codename = new_codename
+                    permission.name = new_name
+                    permission.save(using=self.using)
+                else:
+                    style = color_style()
+                    msg = textwrap.dedent(
+                        f"""
+                        WARNING:
```

Comment:
Before, this silently passed, leaving wrong users & groups attached. Now, there is at least a warning printed. We could adjust this to throw an error in non-interactive mode, so that the "dangerous" thing only happens in development, without necessarily adding an interactive auto-fixer.